app/src/main/java/nl/nfi/cellscanner/MainActivity.java

```diff
@@ -0,0 +1,162 @@
+import android.widget.CompoundButton;
+import android.widget.Switch;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import nl.nfi.cellscanner.recorder.PermissionSupport;
+import nl.nfi.cellscanner.recorder.Recorder;
+
+import static android.support.v4.content.FileProvider.getUriForFile;
+import static nl.nfi.cellscanner.Database.getFileTitle;
+import static nl.nfi.cellscanner.recorder.Recorder.inRecordingState;
+
+public class MainActivity extends AppCompatActivity {
+    /*
+    Activity lifecycle, see: https://developer.android.com/guide/components/activities/activity-lifecycle
+     */
+
+    private Button exportButton, clearButton;
+    private Switch recorderSwitch;
+    private static final int PERMISSION_REQUEST_START_RECORDING = 1;
+
+    /**
+     * Fires when the system first creates the activity
+     * @param savedInstanceState: Bundle object containing the activity's previously saved state
+     */
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        setContentView(nl.nfi.cellscanner.R.layout.activity_main);
+
+        exportButton = findViewById(nl.nfi.cellscanner.R.id.exportButton);
+        clearButton = findViewById(nl.nfi.cellscanner.R.id.clearButton);
+        recorderSwitch = findViewById(nl.nfi.cellscanner.R.id.recorderSwitch);
+        toggleButtonsRecordingState();
+
+        /*
+         Implement checked state listener on the switch that has the ability to start or stop the recording process
+         */
+        recorderSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
+            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
+                if (isChecked) requestStartRecording();
+                else Recorder.stopService(getApplicationContext());
+                toggleButtonsRecordingState();
+            }
+        });
+    }
+
+
+    /**
+     * Request permission to the end user for Location usage. Please be aware that this request is
+     * done on a separate thread
+     */
+    private void requestLocationPermission() {
+            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_COARSE_LOCATION}, PERMISSION_REQUEST_START_RECORDING);
+    }
+
+    /**
+     * Callback function for requests
+     */
+    @Override
+    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
+        switch (requestCode) {
+            case PERMISSION_REQUEST_START_RECORDING: {
+                if (permissions.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                    requestStartRecording();
+                } else {
+                    // explain the app will not be working
+                    Toast.makeText(this,"App will not work without location permissions", Toast.LENGTH_SHORT).show();
+                    toggleButtonsRecordingState();
```

Comment:
Would be convenient to have a new location access requester here; currently will just tell me it doesn't have permissions

Reply:
This is the callback of the request for permission. According to best practice we should not recall here. =D Otherwise we hit an infinite loop

https://developer.android.com/training/permissions/requesting